### PR TITLE
CRM-18483: Refactor for std to support testing & replaying

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
- [CRM-18483: Paypal payments - Pending (Incomplete Transaction)](https://issues.civicrm.org/jira/browse/CRM-18483)

### CODE REVIEW

**Comment by jitendrapurohit** [comment]
Tested the changes, IPN seems to be working fine now.

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit does that mean 'still working fine' or 'working more fine'

(I didn't intentionally fix anything - it was just a refactor - but I did enough to get the test to pass)

**Comment by jitendrapurohit** [comment]
For me, the extern/ipn call failed because the params were retrieved from POST which resulted in an error message : `Failure: Missing Parameter`.

The Pending Contribution which is being created do get completed by externally hitting the url as [mentioned in the issue by Monish](https://issues.civicrm.org/jira/browse/CRM-18483?focusedCommentId=89305&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-89305) including some more params (invoice, mc_gross). 

As these calls failed before this PR, I think it is "working more fine" now ;-). Also, I don't find any more issues related to `Paypal_Standard` after this PR.

File: CRM/Core/Payment/PayPalIPN.php [hunk 1]
```diff
@@ -37,28 +37,42 @@ class CRM_Core_Payment_PayPalIPN extends CRM_Core_Payment_BaseIPN {
...
     static $store = NULL;
-    $value = CRM_Utils_Request::retrieve($name, $type, $store,
-      FALSE, NULL, $location
+    $value = CRM_Utils_Type::validate(
+      CRM_Utils_Array::value($name, $this->_inputParameters),
+      $type,
+      FALSE
     );
     if ($abort && $value === NULL) {
       CRM_Core_Error::debug_log_message("Could not find an entry for $name in $location");
```
Comment (jitendrapurohit) [inline CRM/Core/Payment/PayPalIPN.php]: As this can raise to an accidental notice error, missed to remove `$location` variable here ?

Comment (eileenmcnaughton) [inline CRM/Core/Payment/PayPalIPN.php]: Sure - I just gave you rights to my fork so you can make minor tweaks to my prs

Comment (eileenmcnaughton) [inline CRM/Core/Payment/PayPalIPN.php]: Actually - we shouldn't output a var without escaping - the edit on the line below was a mistake

**Pull request merged**